lang/ruby/lib/avro/json_io.rb

```diff
@@ -0,0 +1,293 @@
+    class JsonDatumReader < DatumReader
+      def validate(schema, data)
+        unless Schema.validate(schema, data)
+          raise AvroTypeError.new(schema, data)
+        end
+      end
+
+      def read_record_data(field, readers_field, decoder)
+        field_data = decoder.data ? decoder.data[field.name] : nil
+        field_decoder = decoder.class.new
+        field_decoder.data = field_data
+        read_data(field.type, readers_field.type, field_decoder)
+      end
+
+      def skip_data(writers_schema, decoder)
+        # no-op
+      end
+
+      def read_record(writers_schema, readers_schema, decoder)
+        record = super
+        validate(readers_schema, record)
+        record
+      end
+
+      def read_fixed(writers_schema, readers_schema, decoder)
+        validate(readers_schema, decoder.data)
+
+        decoder.read_fixed(writers_schema.size)
+      end
+
+      def read_enum(writers_schema, readers_schema, decoder)
+        validate(readers_schema, decoder.data)
+
+        decoder.data
+      end
+
+      def read_array(writers_schema, readers_schema, decoder)
+        read_items = []
+        decoder.data.each do |item|
+          item_decoder = decoder.class.new
+          item_decoder.data = item
+          read_items << read_data(writers_schema.items, readers_schema.items, item_decoder)
+        end if decoder.data
+
+        validate(readers_schema, read_items)
+
+        read_items
+      end
+
+      def read_map(writers_schema, readers_schema, decoder)
+        read_items = {}
+        decoder.data.each_pair do |key, value|
+          value_decoder = decoder.class.new
+          value_decoder.data = value
+          read_items[key] = read_data(writers_schema.values, readers_schema.values, value_decoder)
+        end
+
+        validate(readers_schema, read_items)
+
+        read_items
+      end
+
+      def read_union(writers_schema, readers_schema, decoder)
+        if decoder.data.nil?
+          schema_type, value = :null, nil
+        else
+          schema_type, value = decoder.data.first
+        end
+
+        data_schema = readers_schema.schemas.find do |reader_schema|
+          reader_schema.type_sym == schema_type.to_sym ||
+                                    (reader_schema.name == schema_type if reader_schema.respond_to?(:name))
+        end
+
+        raise AvroTypeError.new(readers_schema, value) unless data_schema
+
+        value_decoder = decoder.class.new
+        value_decoder.data = value
+        field_value = read_data(data_schema, readers_schema, value_decoder)
+
+        validate(data_schema, field_value)
+
+        field_value
+      end
+    end # JsonDatumReader
+
+    # JsonEncoder encodes a given datum and, when asked, writes it to the writer.
+    class JsonEncoder
+      attr_reader :writer
+      attr_accessor :data
+
+      def initialize(writer)
+        @writer = writer
+      end
+
+      def write
+        writer.write(JSON.dump(@data))
```

Comment:
Yeah, you're right about there being a dependency in the current implementation. Looks like the [encoder API in Java](https://github.com/apache/avro/blob/trunk/lang/java/avro/src/main/java/org/apache/avro/io/BinaryDecoder.java#L392) is a bit more general to avoid this. I think it would be great to fix this and then add an encoder/decoder for JSON.